src/frontend/src/scheduler/distributed/stage.rs

```diff
@@ -260,25 +275,66 @@ impl StageRunner {
             .await?;
             return Ok(());
         }
-
-        {
+        // Have been canceled by query manager. No need to send anymore messages.
+        if !handle.is_finished() {
             // Changing state
             let mut s = self.state.write().await;
             match mem::replace(&mut *s, StageState::Failed) {
                 StageState::Started { sender, handle } => {
                     *s = StageState::Running {
-                        _sender: sender,
-                        _handle: handle,
+                        sender: sender,
+                        handle: handle,
                     };
                 }
                 _ => unreachable!(),
             }
+            // All tasks scheduled, send `StageScheduled` event to `QueryRunner`.
+            self.send_event(QueryMessage::Stage(StageEvent::Scheduled(self.stage.id)))
+                .await?;
         }
 
-        // All tasks scheduled, send `StageScheduled` event to `QueryRunner`.
-        self.send_event(QueryMessage::Stage(StageEvent::Scheduled(self.stage.id)))
-            .await?;
+        // Wait on cancel task. So if `run()` is returned. Only if it has been canceled.
+        handle.await?;
+        Ok(())
+    }
 
+    fn cancel_when_required(&mut self) -> SchedulerResult<()> {
```

Comment:
Core idea for this PR:
spawn a task listen on `StageMessage::Stop`. If recevied one, send abort RPC amd change state. 
The stage runner handle is wait on this. Means the handle will only returned if receive a stop message. Can change to return this `cancel_when_required` handle so caller can decide when to await the abort task procedure.